### PR TITLE
Validate speedup

### PR DESCRIPTION
**Pull request by pelszkow**
### Details:
 - *move some method to headers which allow it to be inline*
 - *`PartialShape` c-tors will consume (move) vector of `Dimension`*
 - *initialize map returning by `get_type_info_map` before `main` - purge static function member "mutex"*
 - *find in map and `terminate` if not found `TypeInfo`*

### Tickets:
 - *61261*

### Output:
 - *FakeQuantize::validate_and_infer_types* approximately speedup: 20%

### CODE REVIEW

File: ngraph/core/src/type/element_type.cpp [hunk 1]
```diff
@@ -72,8 +74,20 @@ static const element_types_map_t& get_type_info_map()
...
+
+    const ElementTypes::map_t& get_type_info_map() { return ElementTypes::map; };
+
+    const TypeInfo& get_type_info(element::Type_t type)
+    {
+        const auto& tim = get_type_info_map();
+        const auto& found = tim.find(type);
+        if (found == tim.end())
+        {
+            std::terminate(); // do not support other types
```
Comment (pelszkow) [inline ngraph/core/src/type/element_type.cpp:86]: @ilyachur, I'm not sure if we want such hard `terminate` here but don't see reason to call this functions with `type` which is not one of scope enumeration.

Comment (ilyachur) [reply ngraph/core/src/type/element_type.cpp:86]: Why cannot we throw an exception here?

**Review by ilyachur — COMMENTED** [review]
Do you have any measurements which show improvements?

File: ngraph/core/src/type/element_type.cpp (hunk 1 above)
Comment (pelszkow) [reply ngraph/core/src/type/element_type.cpp:86]: I have thought we have some set of type but after reconsideration `throw` might be better here. So I'll switch to `throw std::out_of_range`.

**Comment by pelszkow** [comment]
> Do you have any measurements which show improvements?
master branch:
```
./unit-test --gtest_filter="IN*smoke_validation*"                                                                                                                                                                                                                           master
Note: Google Test filter = IN*smoke_validation*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from INTERPRETER
[ RUN      ] INTERPRETER.smoke_validation_reshape
[[[[[[ 1.73853e+06us ]]]]]]
[       OK ] INTERPRETER.smoke_validation_reshape (1739 ms)
[ RUN      ] INTERPRETER.smoke_validation_fake_quantize
[[[[[[ 653971us ]]]]]]
[       OK ] INTERPRETER.smoke_validation_fake_quantize (654 ms)
[----------] 2 tests from INTERPRETER (2393 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (2393 ms total)
[  PASSED  ] 2 tests.
```

after this PR:
```
./unit-test --gtest_filter="IN*smoke_validation*"                                                                                                                                                                                                                 validate_speedup
Note: Google Test filter = IN*smoke_validation*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from INTERPRETER
[ RUN      ] INTERPRETER.smoke_validation_reshape
[[[[[[ 1.40525e+06us ]]]]]]
[       OK ] INTERPRETER.smoke_validation_reshape (1405 ms)
[ RUN      ] INTERPRETER.smoke_validation_fake_quantize
[[[[[[ 403533us ]]]]]]
[       OK ] INTERPRETER.smoke_validation_fake_quantize (404 ms)
[----------] 2 tests from INTERPRETER (1809 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (1809 ms total)
[  PASSED  ] 2 tests.
```
Where tests looks like this:
```
diff --git a/ngraph/test/backend/fake_quantize.in.cpp b/ngraph/test/backend/fake_quantize.in.cpp
index 363e89b31..d5bb49490 100644
--- a/ngraph/test/backend/fake_quantize.in.cpp
+++ b/ngraph/test/backend/fake_quantize.in.cpp
@@ -31,6 +31,31 @@ using namespace ngraph;
 static std::string s_manifest = "${MANIFEST}";
 using TestEngine = test::ENGINE_CLASS_NAME(${BACKEND_NAME});

+NGRAPH_TEST(${BACKEND_NAME}, smoke_validation_fake_quantize)
+{
+    const Shape data_shape{1, 2, 3, 4};
+    const size_t levels = 4;
+    const auto data = std::make_shared<op::Parameter>(element::f32, data_shape);
+    const auto input_low = op::Constant::create(element::f32, Shape{}, {0.f});
+    const auto input_high = op::Constant::create(element::f32, Shape{}, {23.f});
+    const auto output_low = op::Constant::create(element::f32, Shape{}, {2.f});
+    const auto output_high = op::Constant::create(element::f32, Shape{}, {16.f});
+
+    const auto quantize = std::make_shared<op::FakeQuantize>(
+        data, input_low, input_high, output_low, output_high, levels);
+
+    const auto start_time = std::chrono::high_resolution_clock::now();
+    constexpr auto validation_rounds = 1000000ul;
+    for (size_t i = 0; i < validation_rounds; ++i)
+    {
+        quantize->validate_and_infer_types();
+    }
+    const float duration = std::chrono::duration_cast<std::chrono::microseconds>(
+                               std::chrono::high_resolution_clock::now() - start_time)
+                               .count();
+    std::cerr << "[[[[[[ " << duration << "us ]]]]]]" << std::endl;
+}
+
 namespace
 {
     template <typename T>
diff --git a/ngraph/test/backend/reshape.in.cpp b/ngraph/test/backend/reshape.in.cpp
index f49870bf1..cb30f54a7 100644
--- a/ngraph/test/backend/reshape.in.cpp
+++ b/ngraph/test/backend/reshape.in.cpp
@@ -32,6 +32,26 @@ static string s_manifest = "${MANIFEST}";

 using TestEngine = test::ENGINE_CLASS_NAME(${BACKEND_NAME});

+NGRAPH_TEST(${BACKEND_NAME}, smoke_validation_reshape)
+{
+    Shape shape_a{2, 2, 3};
+    auto A = make_shared<op::Parameter>(element::f32, shape_a);
+    Shape shape_r{12};
+    auto r = make_shared<op::v1::Reshape>(
+        A, op::Constant::create(element::u64, {shape_r.size()}, shape_r), false);
+
+    const auto start_time = std::chrono::high_resolution_clock::now();
+    constexpr auto validation_rounds = 1000000ul;
+    for (size_t i = 0; i < validation_rounds; ++i)
+    {
+        r->validate_and_infer_types();
+    }
+    const float duration = std::chrono::duration_cast<std::chrono::microseconds>(
+                               std::chrono::high_resolution_clock::now() - start_time)
+                               .count();
+    std::cerr << "[[[[[[ " << duration << "us ]]]]]]" << std::endl;
+}
+
 NGRAPH_TEST(${BACKEND_NAME}, reshape_t2v)
 {
     Shape shape_a{2, 2, 3};
```

**Review by tsocha — APPROVED** [review]

File: ngraph/core/src/partial_shape.cpp [hunk 2]
```diff
@@ -34,15 +34,15 @@ PartialShape::PartialShape(const Shape& shape)
 {
 }
 
-PartialShape::PartialShape(bool rank_is_static, const std::vector<Dimension>& dimensions)
+PartialShape::PartialShape(bool rank_is_static, std::vector<Dimension> dimensions)
     : m_rank_is_static(rank_is_static)
-    , m_dimensions(dimensions)
+    , m_dimensions(std::move(dimensions))
```
Comment (tomdol) [inline ngraph/core/src/partial_shape.cpp:39]: Previously we did a copy in the constructor, now we do a copy before the `dimensions` are passed to the constructor and then perform the move construction. Isn't this less efficient or am I missing something?

File: ngraph/core/src/type/element_type.cpp [hunk 3]
```diff
@@ -12,45 +12,47 @@
...
+    struct ElementTypes
+    {
+        struct element_type_hash
+        {
+            size_t operator()(element::Type_t t) const { return static_cast<size_t>(t); }
+        };
 
-typedef unordered_map<element::Type_t, const TypeInfo, element_type_hash> element_types_map_t;
+        using map_t = std::unordered_map<element::Type_t, TypeInfo, element_type_hash>;
+        static const map_t map;
```
Comment (tomdol) [inline ngraph/core/src/type/element_type.cpp:52]: Just a good-practices suggestion: variable names should not overlap with type names (`std::map` which becomes `map` if someone does `using namespace std`).

**Review by tomdol — APPROVED** [review]

File: ngraph/core/src/partial_shape.cpp (hunk 2 above)
Comment (pelszkow) [reply ngraph/core/src/partial_shape.cpp:39]: If we pass r-value to the constructor we don't create copy before constructor is called. If we want to have supper efficient implementation we need two constructors:
```
PartialShape(const std::vector<Dimension>& dimensions);
PartialShape(std::vector<Dimension>&& dimensions);
```
but in this situation we need to maintain two implementation.

File: ngraph/core/src/type/element_type.cpp (hunk 3 above)
Comment (pelszkow) [reply ngraph/core/src/type/element_type.cpp:52]: Do you suggest I should add some protection for bad practice? :)
...
will do.

Comment (tomdol) [reply ngraph/core/src/type/element_type.cpp:52]: In a way - yes, but using namespace std in .cpp files is quite common.

File: ngraph/core/src/partial_shape.cpp (hunk 2 above)
Comment (pelszkow) [reply ngraph/core/src/partial_shape.cpp:39]: e.g.:
https://github.com/openvinotoolkit/openvino/pull/6779/files#diff-db882ebabc1f33eb7ee439df43f84f81c7170a5e56b9b9c948d9ad28bfdf9b6bR390

**Review by ilyachur — APPROVED** [review]

**Pull request merged**